Fix KTX performance problems

Atlante45 opened the pull request:
Fixes performance issues related to KTX processing on the main thread during progressive loading.

 Performance issues are fixed by offloading the progressive loading of KTX textures to the thread pool.
This required removing direct reads/writes to the NetworkTexture during that processing/file access so as to not use a lock that could cause the same issues.
Instead, I group those operations either before or after the KTX data is processed.

One of the main changes to the code flow is that the initial high mips request used to go through the main mips callback.
I changed it to go through a shared callback with the header data. This makes the code cleaner and easier to follow when cleaning the different requests.

Old code:
KTX processing slows down the main thread
![old profile](https://image.prntscr.com/image/1g28FW8WQ26rxeM25F7NFA.png)

New code:
KTX processing gets offloaded to the thread pool leaving the main thread untouched:
![old profile](https://image.prntscr.com/image/1ghT9U_vS2ul3_8wvzvb-g.png)

Those 2 profiles were made on a cold cache. This still happened on a hot cache but to a lesser effect.

# Test Plan:
- Please report any odd behaviors concerning textures/skyboxes and loading/rendering in general.

## Loading Test:
1) Clear your cache (ex: C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689)
2) Start the Sandbox and then Interface
3) Run through the tutorial (You might need to jump to `localhost/tutorial`)
    - Make sure the text sign are high resolution and readable
    - Make sure all the textures are loaded during the tutorial (No grey objects)
4) Once you are done with the tutorial, make sure the skybox in the garden is rendered with high resolution.
5) Shut down the sandbox
6) Make sure you can see a high resolution for the default skybox
7) Go to Welcome/Zaru and compare what you see to the current stable build.
     - Pay specific attention to: The tree leafs, the blue glow of the portals, and the wooden floor in welcome.
8) Close interface
9) Run through step 2-8 again *without* clearing your cache.

## Retries Test:
1) Start the Sandbox
2) Open Interface
3) Go to `localhost/garden`
4) Close Interface
5) Clear your cache (ex: C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689)
6) Go to the nodes page (http://localhost:40100)
    - You should see a list of connected nodes with a cross on the right side to kill them.
    - The node that interests us is of type "asset-server"
    - We will be killing that server soon by clicking on that kill button
7) Start Interface
    - As soon as you see the first objects starting to load in, kill the asset server
8) In your log, you should notice the following messages:
    `Retryable error while loading http://[some-resource-url] attempt: 1 attemptsRemaining: 7`
9) Look for one of those URLs in your log and make sure it eventually loads correctly.

## Cancel Test:
1) Clear your cache (ex: C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689)
2) Open Interface
3) Open your stats (`/`) and extend them (click on it)
3) Go to Zaru 
4) Close interface when you start seeing `.ktx#head` files being requested
    - Bottom of the third column 
5) Open you last log ( C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689\Interface\Logs\hifi-log.txt)
6) Search your log for a `.ktx#head` request that doesn't have a `.ktx#0` request and save that URL
7) Open Interface
8) Wait for Zaru to fully load
9) Close interface
10) Open you last log ( C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689\Interface\Logs\hifi-log.txt)
11) Make sure the URL you save earlier has a `.ktx#0` request in that log


Some of those steps might be tricky if you are not familiar with the product, especially examining the log.
Don't hesitate to ping me directly here, on Slack or on Gitter for assistance.

@OmegaHeron Would you mind giving this a spin on i5?

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/694/HighFidelity-Beta-PR10689-f6871f6caf83d6a67fb6288d64c970805f667450.exe

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/708/HighFidelity-Beta-PR10689-34df02d29d6f32d37f51169c24780baf7dbe18c6.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/708/HighFidelity-Beta-PR10689-34df02d29d6f32d37f51169c24780baf7dbe18c6.dmg

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/713/HighFidelity-Beta-PR10689-06ab2942ab19f9db382fe9224bb1aa94e7391b0a.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/713/HighFidelity-Beta-PR10689-06ab2942ab19f9db382fe9224bb1aa94e7391b0a.exe

jherico commented on libraries/networking/src/FileCache.cpp:
```diff
@@ -327,6 +331,7 @@ File::~File() {
 
 void File::touch() {
+    PROFILE_RANGE(app, __FUNCTION__)
     utime(_filepath.c_str(), nullptr);
```
I think it would be completely reasonable to push these methods onto another thread, probably like this...

```
void File::touch() {
    std::weak_ptr<File> weakPointer = shared_from_this();
    QtConcurrent::run([weakPointer] {
        auto filePointer = weakPointer.lock();
        utime(filePointer->_filepath.c_str(), nullptr);
        filePointer->_modified = std::max<int64_t>(QFileInfo(filePointer->_filepath.c_str()).lastRead().toMSecsSinceEpoch(), filePointer->_modified);
    });
```

Atlante45 replied:
Somehow I totally missed this comment before.

I like the idea, but I think we'd end up more work than we need too.
For example in the case where you load several textures with different paths but the same hash (Which already happens in Zaru with the couches), we'd end up doing this possible multiple times.

This also made me realize that we should take a hard look at potential concurrency issues here since the file cache and the files could be accessed by several threads at the same time.

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/721/HighFidelity-Beta-PR10689-33adc48d41f26d48cb0ee9752c351a6af19afcc6.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/721/HighFidelity-Beta-PR10689-33adc48d41f26d48cb0ee9752c351a6af19afcc6.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/740/HighFidelity-Beta-PR10689-310dfbab7cb935a30b286d64e39392124622c360.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/740/HighFidelity-Beta-PR10689-310dfbab7cb935a30b286d64e39392124622c360.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/741/HighFidelity-Beta-PR10689-b75aa2486b5963ed09496aa8e04f178ba26d09e1.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/741/HighFidelity-Beta-PR10689-b75aa2486b5963ed09496aa8e04f178ba26d09e1.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/753/HighFidelity-Beta-PR10689-5a5ce79c098cf7255cbcb9d3941ccc8455cfaded.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/753/HighFidelity-Beta-PR10689-5a5ce79c098cf7255cbcb9d3941ccc8455cfaded.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/755/HighFidelity-Beta-PR10689-02e98e963642066b0355d84bfd242ba92a66e8c2.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/755/HighFidelity-Beta-PR10689-02e98e963642066b0355d84bfd242ba92a66e8c2.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/807/HighFidelity-Beta-PR10689-097e75285a94883906be3f54a93c34211fad79ca.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/807/HighFidelity-Beta-PR10689-097e75285a94883906be3f54a93c34211fad79ca.dmg

huffman commented on libraries/model-networking/src/model-networking/TextureCache.cpp:
```diff
@@ -453,235 +452,323 @@ void NetworkTexture::startMipRangeRequest(uint16_t low, uint16_t high) {
+        auto extraInfo = _url == _activeUrl ? "" : QString(", %1").arg(_activeUrl.toDisplayString());
+        qCDebug(networking).noquote() << QString("Request finished for %1%2").arg(_url.toDisplayString(), extraInfo);
+
+         if (_ktxResourceState == REQUESTING_MIP) {
```
Looks like there's an extra space of indentation here

huffman commented on libraries/model-networking/src/model-networking/TextureCache.cpp:
```diff
@@ -453,235 +452,323 @@ void NetworkTexture::startMipRangeRequest(uint16_t low, uint16_t high) {
+            auto self = _self;
+            auto url = _url;
+            auto data = _ktxMipRequest->getData();
+            auto mipLevel = _ktxMipLevelRangeInFlight.first;
```
At some point I think we should just get rid of the range here and use a single number

YuppyPlays commented:
Can test when built :)

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/820/HighFidelity-Beta-PR10689-d090ac4396693426fc08a4d3d40f89ca4f39f3f9.dmg

Atlante45 commented:
test this please

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/826/HighFidelity-Beta-PR10689-d090ac4396693426fc08a4d3d40f89ca4f39f3f9.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/826/HighFidelity-Beta-PR10689-d090ac4396693426fc08a4d3d40f89ca4f39f3f9.dmg

YuppyPlays commented:
Testing.

YuppyPlays commented:
Testing complete:


**Loading Test:**
Clear your cache (ex: C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689) **Check**
Start the Sandbox and then Interface **Check**
Run through the tutorial (You might need to jump to localhost/tutorial) **Check**
Make sure the text sign are high resolution and readable **Check**
Make sure all the textures are loaded during the tutorial (No grey objects) **Check**
Once you are done with the tutorial, make sure the skybox in the garden is rendered with high resolution. **Check**
Shut down the sandbox **Check**
Make sure you can see a high resolution for the default skybox **Check**
Go to Welcome/Zaru and compare what you see to the current stable build. **Check**
Pay specific attention to: The tree leafs, the blue glow of the portals, and the wooden floor in welcome.
Close interface **Check, I noticed in either build for some reason the tree would randomly not load in after clearing cache when retesting? This may be unrelated though**
Run through step 2-8 again without clearing your cache. **Check, Everything ran the same except for the tree not showing up again as in previous step.**


**Retries Test:**

Start the Sandbox **Check**
Open Interface **Check**
Go to localhost/garden **Check**
Close Interface **Check**
Clear your cache (ex: C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689) **Check**
Go to the nodes page (http://localhost:40100) **Check**
You should see a list of connected nodes with a cross on the right side to kill them. **Check**
The node that interests us is of type "asset-server" **Check**
We will be killing that server soon by clicking on that kill button **Check**
Start Interface **Check**
As soon as you see the first objects starting to load in, kill the asset server **Check**
In your log, you should notice the following messages:
Retryable error while loading http://[some-resource-url] attempt: 1 attemptsRemaining: 7 **Check**
Look for one of those URLs in your log and make sure it eventually loads correctly. **Check**


**Cancel Test:**
Clear your cache (ex: C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689) **Check**
Open Interface **Check**
Open your stats (/) and extend them (click on it) **Check**
Go to Zaru **Check**
Close interface when you start seeing .ktx#head files being requested **Check**
Bottom of the third column **Check**
Open you last log ( C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689\Interface\Logs\hifi-log.txt) **Check**
Search your log for a .ktx#head request that doesn't have a .ktx#0 request and save that URL **Going through the log I am not sure if I am searching correctly? Attaching log named hifi-log1 for dev to see**
Open Interface **Check**
Wait for Zaru to fully load **Check**
Close interface **Check**
Open you last log ( C:\Users\USERNAME\AppData\Local\High Fidelity - PR10689\Interface\Logs\hifi-log.txt) **Check**
Make sure the URL you save earlier has a .ktx#0 request in that log **Attaching log named hifi-log2 as I'm not sure I see it**
[hifi-log1.txt](https://github.com/highfidelity/hifi/files/1082362/hifi-log1.txt)
[hifi-log2.txt](https://github.com/highfidelity/hifi/files/1082364/hifi-log2.txt)

Please let me know if there is anymore I can do/provide to assist :)

Atlante45 commented:
Thanks @YuppyPlays, I checked your logs and they are good.

For the trees, that's my bad, it turns out they've been removed from Welcome. They are only in dev-welcome at the moment AFAICT.

jherico commented:
@Atlante45 can you add some commentary to what this PR is doing?

jherico commented on libraries/model-networking/src/model-networking/TextureCache.cpp:
```diff
@@ -453,235 +452,323 @@ void NetworkTexture::startMipRangeRequest(uint16_t low, uint16_t high) {
+    _ktxMipRequest = nullptr;
 }
 
 // This is called when the header or top mips have been loaded
```
the 'or' in this comment should be 'and' now, right?

Atlante45 replied:
Correct, fixing.

OmegaHeron commented:
@Atlante45 - Running i5 CPU comparisons now.  Sorry for delay - I, for some reason, only got notice I had been mentioned in this PR today.

jherico commented on libraries/model-networking/src/model-networking/TextureCache.h:
```diff
@@ -102,16 +101,13 @@ public slots:
     KTXResourceState _ktxResourceState { PENDING_INITIAL_LOAD };
 
-    // TODO Can this be removed?
-    KTXFilePointer _file;
```
I'm working on a PR that also touches this, and one of the problems was that this wasn't populated on all code paths.  The reason this is a member is that it is cache entry lock.  Once the KTXFilePointer for a given cache entry goes completely out of scope the cache is free to eject the file if it grows too large.  Therefore anything that is going to read or write to the file with the expectation that it won't vanish needs to retain a copy of the cache entry.  

There is at least one code path in the existing code that fails to populate _file after a call to `Texture::unserialize`, and in my #10735 PR I fix both that and the fact that the the `gpu::Texture` can briefly outlive the `NetworkTexture` and therefore can crash if the file is ejected from the cache while the GPU backend is buffering it.

Atlante45 replied:
That makes sense, I'll update when your PR is merged and resolve the conflicts.

I see you added a reference the cached file in the KtxStorage class.
So as long as we make sure any KTX texture has a KTX backing, we don't need to store that handle in the NetworkTexture class, correct?

jherico replied:
Yes, I've looked through the code as it is and it looks like there's always a gpu::Texture or KTXFilePointer in scope when writing in this code, so I think we're fine regarding cache behavior.

OmegaHeron commented:
Windows 10 - i5 6500 CPU - GTX970 GPU

Seems fine for CPU use.

Desktop mode 1600x900 maximized window.  Five runs each of last RC (6693) vs this PR (10689).  Each run same location in dev-welcome for 2 minutes, avatar rotating during entire 2 minute test run.  Cache manually deleted between each run.

Blue trace = Master -- Red Trace = This PR

![6693vs10689_cold-cache](https://user-images.githubusercontent.com/8539933/27263753-db19acf2-543d-11e7-84b2-2109e4c1ac0e.jpg)

Same test process as above, except, fully populated cache on each run.

![6693vs10689_hot-cache](https://user-images.githubusercontent.com/8539933/27263934-9194bd7e-5442-11e7-98db-f0a99220a42c.jpg)

Atlante45 commented:
@OmegaHeron No problem, I added you in the PR with an edit. I wonder if somehow that doesn't send a notification in that case.

Either way, thanks for testing!

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/856/HighFidelity-Beta-PR10689-c4c614293598686ed7b4579f096bed8d2e653a05.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/856/HighFidelity-Beta-PR10689-c4c614293598686ed7b4579f096bed8d2e653a05.dmg

Atlante45 commented:
@jherico I updated from the current master and fixed the conflicts there was with your PR.
It hasn't changed in this PR, but I still went through the different code path to create a `gpu::Texture` and they all set a KTX backing.
